### PR TITLE
fix(core): Fix Type<T> for strictFunctionTypes

### LINKED ISSUE #29905: New TestBed.get shows deprecation warning for abstract types and everywhere if using a strict tsconfig
# 🐞 bug report

### Affected Package

The issue is caused by package @angular/core/testing

### Is this a regression?

Yes, the previous version in which this bug was not present was: `8.0.0-beta.11`

### Description

`8.0.0-beta.12` introduced a new `TestBed.get` method and deprecated the existing one (see 
https://github.com/angular/angular/commit/609024f93dc2df038ca5c3d8bfc2406ca5a2a6ac):

https://github.com/angular/angular/blob/609024f93dc2df038ca5c3d8bfc2406ca5a2a6ac/packages/core/testing/src/test_bed.ts#L246-L256

The new signature should be transparent for users as `const service = TestBed.get(UserService);` works.

But the new signature does not work with abstract types, like for example `HttpTestingController` (see https://github.com/angular/angular/blob/master/packages/common/http/testing/src/api.ts#L29) which leads to confusing deprecation warnings.

Update: this is even happening on every `TestBed.get` calls if you have a strict tsconfig, see https://github.com/angular/angular/issues/29905#issuecomment-483651602

## 🔬 Minimal Reproduction

```typescript
import { TestBed } from '@angular/core/testing';
import { HttpClientTestingModule, HttpTestingController } from '@angular/common/http/testing';

describe('UserService', () => {
  beforeEach(() => TestBed.configureTestingModule({
    imports: [HttpClientTestingModule]
  }));

  it('should be created', () => {
    const http: HttpTestingController = TestBed.get(HttpTestingController);
    expect(http).toBeTruthy();
  });
});
```

## 🔥 Exception or Error

`ng lint` shows:

```
/src/app/user.service.spec.ts:10:49 - get is deprecated: from v8.0.0 use Type<T> or InjectionToken<T>
```

## 🌍  Your Environment

**Angular Version:**
<pre><code>
Angular CLI: 8.0.0-beta.13
Node: 11.11.0
OS: darwin x64
Angular: 8.0.0-beta.12
... animations, common, compiler, compiler-cli, core, forms
... language-service, platform-browser, platform-browser-dynamic
... router

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.800.0-beta.13
@angular-devkit/build-angular     0.800.0-beta.13
@angular-devkit/build-optimizer   0.800.0-beta.13
@angular-devkit/build-webpack     0.800.0-beta.13
@angular-devkit/core              8.0.0-beta.13
@angular-devkit/schematics        8.0.0-beta.13
@angular/cli                      8.0.0-beta.13
@ngtools/webpack                  8.0.0-beta.13
@schematics/angular               8.0.0-beta.13
@schematics/update                0.800.0-beta.13
rxjs                              6.4.0
typescript                        3.4.3
webpack                           4.29.6
</code></pre>

cc @Goodwine who wrote it and @mhevery who reviewed: I have not followed closely but couldn't `get<T>(token: T | InjectionToken<T>, notFoundValue?: T, flags?: InjectFlags): any;` be enough?

### PR DESCRIPTION
**Pull request by Goodwine**
See - https://github.com/Microsoft/TypeScript/issues/24428
tl;dr - When using `--strict` mode, `any` isn't a subtype of `string`
and thus the function doesn't match types.

Fixes #29905

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Current definition for `Type<T>` uses `any[]` for the `new` function, however this doesn't work well with `strictFunctionTypes` because for example `new(foo:string)` isn't assignable because `any` isn't a subtype of string.

See - https://github.com/Microsoft/TypeScript/issues/24428

Issue Number: #29905

## What is the new behavior?

`Type<T>` now uses `never[]` which means we can't instantiate with `new` whatever we cast to `Type<T>` but now it should be assignable because `never` is a subtype of `string`

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by Goodwine** [comment]
Ugh, this breaks Angular Material because they have something like:

```ts
const foo = Foo as Type<Foo>;
new foo(bar); // Bar is not assignable to 'never'
```

**Comment by alfaproject** [comment]
Why not just make a breaking change and have the proper type? Or wasn't it deprecated for long enough?

**Comment by Goodwine** [comment]
@alfaproject I totally agree, however the following things happened:

- `Injector.get` was deprecated and it will break in the next major release, yes. :crossed_fingers: 
- `TestBed.get` wasn't deprecated the PR that caused this bug was the one that added `@deprecated` to the signature that takes `any`
- but then `--strict` makes it so that any class that has a constructor with >1 parameters doesn't match the better-typed constructor (which still returns `any`) and instead matches the deprecated signature and casting to `Type<T>` doesn't work
- I fixed `Type<T>` (this PR) to work with `--strict` but then it breaks uses similar to the one described on my earlier comment about Angular Material

**Comment by Goodwine** [comment]
I'll send another PR with hopefully a better plan

**Comment by Goodwine** [comment]
New PR with different approach -
https://github.com/angular/angular/pull/30036
